### ISSUE TITLE
Remove Qt dependency in AutoGTP(?)

### ISSUE THREAD

**Issue by apetresc** [gcp/leela-zero#1550]
Are we philosophically opposed to removing Qt from autogtp? I understand it's convenient, but it seems insane to pull in a >3GB set of dependencies to essentially simplify some string operations and threading primitives.

If I put together a PR that reimplements Qt functionality in autogtp (with the understanding that it might be slightly less elegant) would it be dropped on philosophical grounds, or is this open to debate?

**Comment by bood** [comment]
I don't see why Qt is a problem, it's a great library. (Is it really 3GB? when I install the Qt dependency, it's pretty fast)

BTW If you really want to do something, maybe re-writing it in Rust? I have plan re-writing the engine part with Rust as a learning project myself. :-)

**Comment by alreadydone** [comment]
I can confirm that Qt occupies 3.87 GB on my Windows machine. I think it's a minimal installation just for it to work with VS2017 (I remember I only selected MSVC related stuff to install), but maybe it can be cut down further. I wasn't even trying to compile `autogtp`, only `validation`.

**Comment by l1t1** [comment]
i searched qt in the source code and got result as following

[qt_find.zip](https://github.com/gcp/leela-zero/files/2096429/qt_find.zip)

**Comment by zhanzhenzhen** [comment]
+1 for removing Qt. For me it's not related to size. I don't want to install many things in my Mac. If it's something that can be directly put to "/usr/local", that's fine. But if it must be "installed", that's evil to me.

**Comment by l1t1** [comment]
is it possible put all needed qt source file in leelaz and not install qt?

**Comment by marcocalignano** [comment]
Are you sure you can maintain portability between 3 different operating systems without having to write `#IF WIN32` ... etc? Autogtp is heavily multithreading/multitasking program. You will need to test it for every platform.

**Comment by ihavnoid** [comment]
Hmm, I think a lot of the required cross-platform libraries can be obtained from boost?  We need boost for leelaz anyway...

**Comment by marcocalignano** [comment]
I do not know It seems a lot work, a lot of time, a ot of testing to change a running system just to save 3GB. I would first look at removing Qt DLL that you do not need (like QtGui QtWidget all the styles libs).

It is like when your car does not fit in your garage anymore and you start building a new smaller car instead of trying taking out all the junk from your garage first.

**Comment by ihavnoid** [comment]
> I do not know It seems a lot work, a lot of time, a ot of testing to change a running system just to save 3GB. I would first look at removing Qt DLL that you do not need (like QtGui QtWidget all the styles libs).

I haven't given a lot of thought on how much work it might take.  My experience on my prior (professional) work is that re-inventing the wheel is much more expensive than originally thought, so I do have concerns on how hard it is and how long it will take.  However, this is an open-source project which relies on each individual's spare time, and if somebody is well motivated enough to spend their spare time implementing it, I do not see any reason to refuse contribution as long as the resulting code is feature-complete, robust and stable.

**Comment by marcocalignano** [comment]
Yes but since time (of competent programmers) is the main currency for an open source project, I believe that this currency could be spent better elsewhere.

**Comment by ihavnoid** [comment]
@marcocalignano - Maybe, maybe not.  It's not like somebody who is interested in contributing on one thing is willing to spend time contributing on something else.  In other words, the 'currency' isn't really exchangeable.

If this was a commercial project then yes I would oppose spending time rewriting autogtp for the very same reason that you mentioned.

**Comment by l1t1** [comment]
as a common windows user, i think the code is not very important since we have binary file. and more, qt only be used in autogtp and validation, to a person who want selfplay but not upload to the project, use gogui instead.

**Comment by odeint** [comment]
>  i think the code is not very important since we have binary file. and more, qt only be used in autogtp and validation

This is an important point. Even if you are interested in compiling your own LeelaZero engine, you can still use validation and autogtp from a binary release. It is only when you want to change things in autogtp itself that you need Qt. Not really a pressing issue, but of course if someone really wants to make a autogtp_QtFree version as a project, why not!

**Comment by joseph-ireland** [comment]
The qt installer for windows is crazy! Installing everything on linux with something like `yum install qt5*` only takes up like 200mb, and installing the basics with `yum install qt5-qtbase-devel` only takes up like 40mb, containing everything you need for this. 

The windows installer, on the other hand, insists on installing 4 copies of every single library: a debug and release version of both static and dynamic libraries, along with all the debug info for both. On top of that, it forces you to install Qt Creator which has it's own 5th copy of qt libs. All of this is compulsory, with no way to opt out in the installer, producing a running total of 4GB. For a project like this, which only needs that 40MB install, it's just ridiculous.

**Comment by l1t1** [comment]
katago removed boost dependency  now https://github.com/lightvector/KataGo/pull/335#issuecomment-723761918
